File: go/core/parallel_tools_test.go
package core

import (
	"context"
	"strings"
	"sync"
	"testing"
	"time"
)

// Tests for concurrent (parallel) tool-call execution. When
// AgentOptions.ParallelToolCalls is true and an assistant turn returns >=2 tool
// calls, dispatches run concurrently (goroutines + sync.WaitGroup) — but the
// tool-result messages must still be appended in the original ToolCalls order so
// the transcript is deterministic. Default (false) keeps sequential dispatch.

// multiToolCall builds an assistant response requesting several tool calls at once.
func multiToolCall(calls ...ToolCall) ChatResponse {
	return ChatResponse{ToolCalls: calls}
}

func toolResults(messages []ChatMessage) []string {
	var out []string
	for _, m := range messages {
		if m.Role == "tool" {
			out = append(out, m.Content)
		}
	}
	return out
}

func TestParallelDispatchOverlaps(t *testing.T) {
	// Two tools that each block until both have started — only completes if concurrent.
	var mu sync.Mutex
	started := 0
	bothStarted := make(chan struct{})
	slow := func(name string) Tool {
		return FuncTool{ToolName: name, Params: map[string]any{"type": "object"},
			Fn: func(ctx context.Context, _ map[string]any) (string, error) {
				mu.Lock()
				started++
				if started == 2 {
					close(bothStarted)
				}
				mu.Unlock()
				<-bothStarted
				return name, nil
			}}
	}
	mock := NewMockLlmProvider()
	mock.PushResponse(multiToolCall(ToolCall{ID: "c1", Name: "a"}, ToolCall{ID: "c2", Name: "b"})).PushText("done")
	agent := NewSmoothAgent(mock, AgentOptions{Tools: []Tool{slow("a"), slow("b")}, ParallelToolCalls: true})

	done := make(chan AgentRunResponse, 1)
	go func() {
		res, err := agent.Run(context.Background(), "go", nil)
		if err != nil {
			t.Error(err)
		}
		done <- res
	}()
	select {
	case res := <-done:
		if res.Text != "done" || res.ToolCalls != 2 {
			t.Fatalf("unexpected result: %+v", res)
		}
	case <-time.After(3 * time.Second):
		t.Fatal("timed out — tools did not run concurrently")
	}
}

func TestParallelOrderPreserved(t *testing.T) {
	gates := map[string]chan struct{}{"A": make(chan struct{}), "B": make(chan struct{}), "C": make(chan struct{})}
	make1 := func(name string) Tool {
		return FuncTool{ToolName: name, Params: map[string]any{"type": "object"},
			Fn: func(ctx context.Context, _ map[string]any) (string, error) {
				<-gates[name]
				return "result-" + name, nil
			}}
	}
	mock := NewMockLlmProvider()
	mock.PushResponse(multiToolCall(ToolCall{ID: "c1", Name: "A"}, ToolCall{ID: "c2", Name: "B"}, ToolCall{ID: "c3", Name: "C"})).PushText("done")
	agent := NewSmoothAgent(mock, AgentOptions{Tools: []Tool{make1("A"), make1("B"), make1("C")}, ParallelToolCalls: true})

	done := make(chan struct{})
	go func() {
		if _, err := agent.Run(context.Background(), "go", nil); err != nil {
			t.Error(err)
		}
		close(done)
	}()
	// Finish in B, C, A order — opposite of transcript order for A.
	time.Sleep(5 * time.Millisecond)
	close(gates["B"])
	time.Sleep(5 * time.Millisecond)
	close(gates["C"])
	time.Sleep(5 * time.Millisecond)
	close(gates["A"])
	<-done

	got := toolResults(mock.Calls()[1].Messages)
	want := []string{"result-A", "result-B", "result-C"}
	if len(got) != 3 || got[0] != want[0] || got[1] != want[1] || got[2] != want[2] {
		t.Fatalf("tool results out of order: got %v want %v", got, want)
	}
}

func TestParallelFailingToolKeepsPosition(t *testing.T) {
	ok := func(name string) Tool {
		return FuncTool{ToolName: name, Params: map[string]any{"type": "object"},
			Fn: func(context.Context, map[string]any) (string, error) { return "ok", nil }}
	}
	boom := FuncTool{ToolName: "B", Params: map[string]any{"type": "object"},
		Fn: func(context.Context, map[string]any) (string, error) { return "", &kaboomErr{} }}
	mock := NewMockLlmProvider()
	mock.PushResponse(multiToolCall(ToolCall{ID: "c1", Name: "A"}, ToolCall{ID: "c2", Name: "B"}, ToolCall{ID: "c3", Name: "C"})).PushText("done")
	agent := NewSmoothAgent(mock, AgentOptions{Tools: []Tool{ok("A"), boom, ok("C")}, ParallelToolCalls: true})
	if _, err := agent.Run(context.Background(), "go", nil); err != nil {
		t.Fatal(err)
	}
	got := toolResults(mock.Calls()[1].Messages)
	if len(got) != 3 || got[0] != "ok" || !strings.Contains(got[1], "kaboom") || got[2] != "ok" {
		t.Fatalf("unexpected tool results: %v", got)
	}
}

type kaboomErr struct{}

func (kaboomErr) Error() string { return "kaboom" }

func TestParallelDefaultOffSequential(t *testing.T) {
	var order []string
	var mu sync.Mutex
	make1 := func(name string) Tool {
		return FuncTool{ToolName: name, Params: map[string]any{"type": "object"},
			Fn: func(context.Context, map[string]any) (string, error) {
				mu.Lock()
				order = append(order, name)
				mu.Unlock()
				return name, nil
			}}
	}
	mock := NewMockLlmProvider()
	mock.PushResponse(multiToolCall(ToolCall{ID: "c1", Name: "A"}, ToolCall{ID: "c2", Name: "B"})).PushText("done")
	agent := NewSmoothAgent(mock, AgentOptions{Tools: []Tool{make1("A"), make1("B")}}) // ParallelToolCalls defaults false
	res, err := agent.Run(context.Background(), "go", nil)
	if err != nil {
		t.Fatal(err)
	}
	if len(order) != 2 || order[0] != "A" || order[1] != "B" {
		t.Fatalf("sequential dispatch order wrong: %v", order)
	}
	if res.ToolCalls != 2 {
		t.Fatalf("want 2 tool calls, got %d", res.ToolCalls)
	}
}

func TestParallelSingleToolCallIdentical(t *testing.T) {
	echo := FuncTool{ToolName: "echo", Params: map[string]any{"type": "object"},
		Fn: func(_ context.Context, args map[string]any) (string, error) {
			s, _ := args["text"].(string)
			return s, nil
		}}
	mock := NewMockLlmProvider()
	mock.PushToolCall("c1", "echo", `{"text":"hi"}`).PushText("done")
	agent := NewSmoothAgent(mock, AgentOptions{Tools: []Tool{echo}, ParallelToolCalls: true})
	res, err := agent.Run(context.Background(), "go", nil)
	if err != nil {
		t.Fatal(err)
	}
	if res.Text != "done" || res.ToolCalls != 1 {
		t.Fatalf("unexpected result: %+v", res)
	}
}

File: typescript/core/test/parallelTools.test.ts
/**
 * Unit tests for concurrent (parallel) tool-call execution.
 *
 * When `AgentOptions.parallelToolCalls` is true and an assistant turn returns ≥2
 * tool calls, the dispatches run concurrently (`Promise.all`) — but the tool-result
 * messages must still be appended in the original `tool_calls` order so the
 * transcript is deterministic. Default (false) keeps sequential dispatch.
 */

import { describe, expect, it } from 'vitest';
import { Tool } from '../src/agent.js';
import { SmoothAgent } from '../src/agent.js';
import { MockLlmProvider, ScriptedMessage } from '../src/llmProvider.js';

/** An OpenAI-shaped assistant message requesting several tool calls at once. */
function multiToolCall(...calls: Array<[string, string, string]>): ScriptedMessage {
    return {
        content: null,
        tool_calls: calls.map(([id, name, args]) => ({ id, function: { name, arguments: args } })),
    };
}

/** A deferred promise + resolve handle. */
function gate(): { promise: Promise<void>; release: () => void } {
    let release!: () => void;
    const promise = new Promise<void>((resolve) => {
        release = resolve;
    });
    return { promise, release };
}

describe('parallel tool calls', () => {
    it('overlaps tool dispatches when enabled', async () => {
        // Two tools that each block until both have started — only completes if concurrent.
        let started = 0;
        const bothStarted = gate();
        const slow: (name: string) => Tool = (name) => ({
            name,
            description: '',
            parameters: { type: 'object' },
            execute: async () => {
                started++;
                if (started === 2) bothStarted.release();
                await bothStarted.promise;
                return name;
            },
        });
        const client = new MockLlmProvider();
        client.pushResponse(multiToolCall(['c1', 'a', '{}'], ['c2', 'b', '{}'])).pushText('done');
        const agent = new SmoothAgent(client, { tools: [slow('a'), slow('b')], parallelToolCalls: true });
        const result = await agent.run('go');
        expect(result.text).toBe('done');
        expect(result.toolCalls).toBe(2);
    });

    it('preserves tool-result order despite scrambled completion', async () => {
        const gates: Record<string, ReturnType<typeof gate>> = { A: gate(), B: gate(), C: gate() };
        const make: (name: string) => Tool = (name) => ({
            name,
            description: '',
            parameters: { type: 'object' },
            execute: async () => {
                await gates[name].promise;
                return `result-${name}`;
            },
        });
        const client = new MockLlmProvider();
        client.pushResponse(multiToolCall(['c1', 'A', '{}'], ['c2', 'B', '{}'], ['c3', 'C', '{}'])).pushText('done');
        const agent = new SmoothAgent(client, { tools: [make('A'), make('B'), make('C')], parallelToolCalls: true });

        const run = agent.run('go');
        // Finish in B, C, A order — opposite of transcript order for A.
        await new Promise((r) => setTimeout(r, 5));
        gates.B.release();
        await new Promise((r) => setTimeout(r, 5));
        gates.C.release();
        await new Promise((r) => setTimeout(r, 5));
        gates.A.release();
        await run;

        const toolResults = client.calls[1].messages.filter((m) => m.role === 'tool').map((m) => m.content);
        expect(toolResults).toEqual(['result-A', 'result-B', 'result-C']);
    });

    it('keeps a failing tool in its correct position', async () => {
        const ok: (name: string) => Tool = (name) => ({
            name,
            description: '',
            parameters: { type: 'object' },
            execute: async () => 'ok',
        });
        const boom: Tool = {
            name: 'B',
            description: '',
            parameters: { type: 'object' },
            execute: async () => {
                throw new Error('kaboom');
            },
        };
        const client = new MockLlmProvider();
        client.pushResponse(multiToolCall(['c1', 'A', '{}'], ['c2', 'B', '{}'], ['c3', 'C', '{}'])).pushText('done');
        const agent = new SmoothAgent(client, { tools: [ok('A'), boom, ok('C')], parallelToolCalls: true });
        await agent.run('go');

        const toolResults = client.calls[1].messages.filter((m) => m.role === 'tool').map((m) => String(m.content));
        expect(toolResults[0]).toBe('ok');
        expect(toolResults[1]).toContain('kaboom');
        expect(toolResults[2]).toBe('ok');
    });

    it('dispatches sequentially when the flag is off (default)', async () => {
        const order: string[] = [];
        const make: (name: string) => Tool = (name) => ({
            name,
            description: '',
            parameters: { type: 'object' },
            execute: async () => {
                order.push(name);
                return name;
            },
        });
        const client = new MockLlmProvider();
        client.pushResponse(multiToolCall(['c1', 'A', '{}'], ['c2', 'B', '{}'])).pushText('done');
        const agent = new SmoothAgent(client, { tools: [make('A'), make('B')] }); // parallelToolCalls defaults false
        const result = await agent.run('go');
        expect(order).toEqual(['A', 'B']);
        expect(result.toolCalls).toBe(2);
    });

    it('handles a single tool call identically with the flag on', async () => {
        const echo: Tool = {
            name: 'echo',
            description: '',
            parameters: { type: 'object' },
            execute: async (args) => String(args.text ?? ''),
        };
        const client = new MockLlmProvider();
        client.pushToolCall('c1', 'echo', '{"text":"hi"}').pushText('done');
        const agent = new SmoothAgent(client, { tools: [echo], parallelToolCalls: true });
        const result = await agent.run('go');
        expect(result.text).toBe('done');
        expect(result.toolCalls).toBe(1);
    });
});
